Windows: Adds Amcache plugin

dgmcdona opened the pull request:
This adds a plugin for parsing information about installed executables and applications from the `Amcache.hve` registry hive. 

The original volatility plugin supported parsing of `Amcache.hve` hives on Windows 8 systems; this PR includes that functionality, as well as new support for Windows 10+ systems. 

On both Windows 8 and Windows 10 systems, there are analogous 'program' and 'file' type entries. These can be found under the following registry keys:
- Windows 8: `Programs`/`File`
- Windows 10: `InventoryApplication`/`InventoryApplicationFile`

In both cases, `File` entries can be correlated with `Program` entries via a `ProgramId` string. Because a single `Program` can have multiple `files` associated with its installation, this plugin renders them in a tree format, with a `Program` entry being the parent of the `File` entries when the relationship can be determined. Otherwise, `File` entries are treated as standalone and rendered at the top level of the tree. 

It makes sense that there would be standalone `File` entries, since the scheduled task that populates the Amcache periodically scans a fixed set of folders for any PE files, and adds them to the Amcache when found.

On Windows 10 systems, there is a third registry key (`Root\\InventoryDriverBinary`) which contains records of installed drivers. These entries are parsed and rendered at the top level as well.

Where possible, `Program` and `File` entries share fields, and in cases where a field may only apply to one or the other, a `renderers.NotApplicableValue` is used to populate that field.

dgmcdona committed:
Windows: Adds Amcache plugin

dgmcdona commented on volatility3/framework/plugins/windows/amcache.py:
```diff
@@ -0,0 +1,605 @@
+                    amcache.get_key("Root\\InventoryDriverBinary")  # type: ignore
+                )
+            )
+        except KeyError:
```
fixed

dgmcdona commented on volatility3/framework/plugins/windows/amcache.py:
```diff
@@ -0,0 +1,605 @@
+                    program_entry = programs.pop(program_id.strip().strip("\u0000"))
+                    yield (0, program_entry)
+                    files_indent = 1
+                except KeyError:
```
fixed

atcuno commented:
@ikelos can this one get reviewed before it gets stale?

ikelos commented on volatility3/framework/plugins/windows/amcache.py:
```diff
@@ -0,0 +1,603 @@
+    expected_type: reg_extensions.RegValueTypes,
+) -> Any:
+    try:
+        value = next(
```
So I can't quite see why this is using next rather than just a 0 index?  It looks like you're just using the head of the list because every path from this next returns (meaning you never get to the second item)?  I don't specifically have anything against it, but it makes me wonder what's going on (and why it's provided a list of values) if it only ever uses the first one?  I'd have gone with a list comprehension and a 0 index (if the list has any contents)...

dgmcdona replied:
The advantage of `next` with a generator comprehension over using a list comprehension with a zero-index is that creating a complete list of values, even a filtered one, using a list comprehension requires performing the `.get_name().lower().strip().strip("\u0000")` operations on every value, even those in the iterator beyond the one we're looking for. Using `next` with a generator expression only performs those operations as needed, and stops evaluating once the desired value is found.

ikelos replied:
You're right about the efficiency.  I still find standalone calls to `next` a little unusual (rather than a `for...: break`) but that's ok, it just threw me a little (and made me wonder why we stopped after the first one, what are the others used for, but a for loop would have the same effect))?

ikelos commented on volatility3/framework/plugins/windows/amcache.py:
```diff
@@ -0,0 +1,603 @@
+    return value
+
+
+def _sort_entries(entry_tuple):
```
Feels as though this could be a lambda, but again not a show stopper, and it offers clarity, but if it's gonna be a new function then proper typing (and mabye even a docstring) please.  Just because it's short doesn't mean it needs no explaining...

dgmcdona replied:
I split it out into a new function since it's essential that the implementations match for `group_by` and `sorted`, added a comment to it plus type hints.

ikelos commented on volatility3/framework/plugins/windows/amcache.py:
```diff
@@ -0,0 +1,603 @@
+            if isinstance(compile_time, datetime.datetime):
+                yield f"Amcache: {entry_type} {path} compiled (PE metadata)", timeliner.TimeLinerType.MODIFIED, compile_time
+
+    @classmethod
```
This is usually the first method, but I'm not gonna be picky...

ikelos commented on volatility3/framework/plugins/windows/amcache.py:
```diff
@@ -0,0 +1,603 @@
+        kernel: interfaces.context.ModuleInterface,
+    ) -> Optional[registry.RegistryHive]:
+        """Retrieves the `Amcache.hve` registry hive from the kernel module, if it can be located."""
+        return next(
```
Again with the next.  Happy you're consistent but it still makes me look around for a list that doesn't seem to exist...  5:P

ikelos commented on volatility3/framework/plugins/windows/amcache.py:
```diff
@@ -0,0 +1,603 @@
+                last_mod_2_raw,
+                create_time_raw,
+            ) = (
+                _try_get_value(values, "100", reg_extensions.RegValueTypes.REG_SZ),
```
Are these values and their meanings documented somewhere?  Might be nice to have an enum or a dictionary just for humans to translate what the number is to what the value means.  Just feels like like ordered lists of specific numbers like this are fragile, and keeping a separate clear table somewhere makes it easier to tinker or change without breaking everything like a giant game of jenga...  5;)

dgmcdona replied:
I've created enums for each set of Amcache entry value value names, and added a reference in a comment for those who want to seek out more information

ikelos commented on volatility3/framework/plugins/windows/amcache.py:
```diff
@@ -0,0 +1,603 @@
+            )
+
+            create_time = (
+                conversion.wintime_to_datetime(create_time_raw)
```
I feel like we should now have type hints that do this (eg, HexOrAbsent), but reasonably ok with this.  Just wondering if there's someway to make them more compact or shrink the duplication, but I'm not convinced there is save for a full subfunction, so this is probably fine.

ikelos reviewed (commented):
Ok, generally looks ok, a bunch of comments but most of them are stylistic.  The most important one from my perspective would be to get those arbitrary number keys looked up from a dictionary or an enum somewhere, just so that mere humans can see what's supposed to be going on...

dgmcdona commented:
@ikelos I made some additions to the `CM_KEY_VALUE` class to hopefully help future plugin authors avoid common pitfalls when dealing with the registry. The `get_string_value`/`get_binary_value`/`get_int_value` methods are meant to:
- Remind plugin authors that registry value data may not be of the type that they would typically expect when retrieving it, and they should be prepared to handle that case. They have the option of providing a default value, or handling the new exception class themselves when it is raised. The new exception contains the actual value type of the data that they are attempting to access.
- Gracefully handle the two major failure cases. These are that the value may be paged out or otherwise absent, in which case those methods return `None`, or that the value is not of the type they expect, in which case an `UnexpectedValueTypeException` is raised or their supplied default value is returned. In my case, I have set the default to `renderers.UnparsableValue()` since there's nothing else I can do with the data at that point.

It worked well in this case, but I'm open to any suggestions you might have for how to improve this interface.

ikelos commented on volatility3/framework/symbols/windows/extensions/registry.py:
```diff
@@ -5,9 +5,9 @@
+from typing import Iterable, List, Optional, Union
 
-from volatility3.framework import constants, exceptions, interfaces, objects
+from volatility3.framework import T, constants, exceptions, interfaces, objects
```
Please define a new `TypeVar` here, there seems little reason to reuse the only from the framework?  Its purpose is to make sure that the input class and the output class of a function are the same type.

ikelos commented on volatility3/framework/symbols/windows/extensions/registry.py:
```diff
@@ -52,6 +75,24 @@ class RegKeyFlags(enum.IntEnum):
     KEY_VIRTUAL_STORE = 0x200
 
 
+class UnexpectedValueTypeException(Exception):
```
I wondering if this should live here or in `exceptions`?  Which do you think?  Either way I suspect it should probably inherit from `VolatilityException` rather than just a standard `Exception`.

ikelos commented on volatility3/framework/symbols/windows/extensions/registry.py:
```diff
@@ -132,6 +173,103 @@ def get_full_key_name(self) -> str:
+        ]:
+            return data
+        if self.get_type() == RegValueTypes.REG_NONE:
+            return b""
```
So, should we be returning an empty (binary) string here, rather than an actual python `None`?  Do we use `None` as an error condition in this?  Generally empty strings meaning "nothing" are hard to distinguish from strings that are intentionally empty, so I'd prefer to return None if we could, but it's there's a good reason I'll listen to it...

dgmcdona replied:
I'm used to seeing `REG_NONE` values being treated as empty binary containers, so I chose this return type as it was just consistent with my prior experience; I just did a sanity check across a couple of test hives to make sure that this will always be the case. I think we can safely return `None` here, I'll make the change.

ikelos commented on volatility3/framework/symbols/windows/extensions/registry.py:
```diff
@@ -132,6 +173,103 @@ def get_full_key_name(self) -> str:
+        """Properly decodes the data associated with the value node"""
+        # Determine if the data is stored inline
+        datalen = self.DataLength
+        data = b""
```
See later on where I talk about blank strings.

ikelos commented on volatility3/framework/symbols/windows/extensions/registry.py:
```diff
@@ -152,9 +290,173 @@ def get_subkeys(self) -> Iterable[interfaces.objects.ObjectInterface]:
             yield from self._get_subkeys_recursive(hive, subkey_node)
 
+    def get_int_value(
+        self, name: str, default: Optional[T] = None
```
Should the default also have to an `int` (or an `Optional[int]`) here?

dgmcdona replied:
The idea behind using `T` here instead of int was that it gives consumers the ability to decide what type they want to return in the case of missing values. I definitely didn't want `renderers` this deep in the code, so this was my attempt to solve that same problem again.

ikelos commented on volatility3/framework/symbols/windows/extensions/registry.py:
```diff
@@ -152,9 +290,173 @@ def get_subkeys(self) -> Iterable[interfaces.objects.ObjectInterface]:
+                return default
+            raise
+
+        return value if isinstance(value, int) else None
```
When would `get_value_data(whatever, INTEGER_TYPES)` return something that wasn't an integer (or None)?  I really get the feeling it should be possible to return `None`?

ikelos commented on volatility3/framework/symbols/windows/extensions/registry.py:
```diff
@@ -152,9 +290,173 @@ def get_subkeys(self) -> Iterable[interfaces.objects.ObjectInterface]:
+        return value if isinstance(value, str) else None
+
+    def get_multi_string_value(
+        self, name: str, default: Optional[T] = None
```
So, default doesn't have to be a `List` here?

ikelos commented on volatility3/framework/symbols/windows/extensions/registry.py:
```diff
@@ -152,9 +290,173 @@ def get_subkeys(self) -> Iterable[interfaces.objects.ObjectInterface]:
+
+        return (
+            value
+            if isinstance(value, list) and all(isinstance(val, str) for val in value)
```
Don't list iterators always need to be in square brackets?  Fine if not, just stood out to me a little is all.  5:)

ikelos commented on volatility3/framework/symbols/windows/extensions/registry.py:
```diff
@@ -152,9 +290,173 @@ def get_subkeys(self) -> Iterable[interfaces.objects.ObjectInterface]:
+                (
+                    val
+                    for val in self.get_values()
+                    if val.get_name().lower().strip().strip("\u0000") == name.lower()
```
Is case definitely not important?  Presumably you could theoretically get two values that differ only in case?  Just want to check that this is absolutely valid for what we're doing?

dgmcdona replied:
Yeah, registry value names are case insensitive, so this is the behavior we want.

ikelos commented on volatility3/framework/symbols/windows/extensions/registry.py:
```diff
@@ -170,7 +472,7 @@ def _get_subkeys_recursive(
             listjump = 2
         elif node.vol.type_name.endswith(constants.BANG + "_CM_KEY_NODE"):
-            yield node
+            yield node  # type: ignore
```
I'm not a big fan of `# type: ignore` because it often means "we don't want to think about this" and quickly becomes "we've forgotten about this".  What are the circumstances here where we're not yield the right type?

dgmcdona replied:
I just used it there because my language server was complaining about `ObjectInterface` not being an instance of `CM_KEY_NODE`, since it doesn't understand that it's going to have that extension class created instead. I'll remove this and use a `cast` for clarity. We've talked about too much casting being overkill in the past, so if you feel like that's the case here, I'm happy to remove both the cast and the comment.

ikelos replied:
Well, ignoring is definitely bad, but casting feels like somewhat ignoring the problem too, but I guess we'll go with cast.

ikelos commented on volatility3/framework/plugins/windows/amcache.py:
```diff
@@ -0,0 +1,690 @@
+        for field in dataclasses.fields(self):
+            if getattr(self, field.name) is None:
+                # Map 'None' values that have been explicitly set to `NotAvailableValue()`
+                setattr(self, field.name, renderers.NotAvailableValue())
```
I worry that this is using renderer values (which are typically only for the renderer to figure out how to display them) in code that may be used by plugins (thus requiring the plugins to have to deal with them too).  Generally these values are only supplanted in over None values just before they're yielded into TreeGrid.  I'm happy that this is a separate class, and its only purpose may be for displaying output, but if it's like to get used in normal python code, then it may be better to stick to python's `None` type, and then making the yield statement, something like `this or renderers.NotAvailableValue`?

ikelos commented on volatility3/framework/plugins/windows/amcache.py:
```diff
@@ -0,0 +1,690 @@
+    invocations.
+    """
+    program_id, _ = entry_tuple
+    key = program_id if isinstance(program_id, str) else ""
```
Why are we returning a blank string on what essentially looks like failure?  Shouldn't we be distinguishing it somehow (such as returning `None`)?

dgmcdona replied:
`None` isn't comparable to `str` values, and this is being used as a sort key, so we have to return a `str`.

ikelos commented on volatility3/framework/plugins/windows/amcache.py:
```diff
@@ -0,0 +1,690 @@
+                conversion.wintime_to_datetime(value)
+                if isinstance(value, int)
+                else (
+                    renderers.NotAvailableValue()
```
Ok, distinguishing between these two states does seems worthwhile, so I'm happy to be easier on using them/passing them around.

dgmcdona replied:
I was trying to find a good way to differentiate on the things that are 'definitively missing' vs things that have an unexpected value type or could not be recovered due to `InvalidPagedAddressException`s and I'm still not 100% sure I'm happy with this since, as you said, it feels like it would be best to let the consumers of these things deal with the renderers.

ikelos reviewed (commented):
Cool, seems good.  I'm somewhat willing to relent on my suggestion of keeping the AbsentValues until the rendering because you do actually seem to be distinguishing on types.  I still thing you could leave out the default argument and then just provide `value or UnparsableValue(),` in the yield statement, but if the default parameter is useful I won't complain.  Just a couple little house keeping bits, but generally a very clean and tidy plugin with very few deviations for the normal coding style for volatility.  Good work!  5:D

ikelos commented on volatility3/framework/plugins/windows/amcache.py:
```diff
@@ -0,0 +1,690 @@
+# This file is Copyright 2019 Volatility Foundation and licensed under the Volatility Software License 1.0
```
The copyright should be the year the code was first added, so 2024 please.

dgmcdona replied:
Fixed

ikelos commented on volatility3/framework/plugins/windows/amcache.py:
```diff
@@ -0,0 +1,690 @@
+import datetime
+import enum
+import logging
+from itertools import groupby, chain
```
Please just import `itertools` and then reference the methods as `itertools.groupby` or `groupby.chain`.  This prevents people trying to import it from this module rather than where it's actually defined.

dgmcdona replied:
Fixed

ikelos commented on volatility3/framework/symbols/windows/extensions/registry.py:
```diff
@@ -132,6 +175,103 @@ def get_full_key_name(self) -> str:
+        ]:
+            return data
+        if self.get_type() == RegValueTypes.REG_NONE:
+            return None
```
Returning None rather than `b""` is better, but unfortunately is also an API change, so strictly this needs a the MAJOR version bumping (which I suspect will be of the whole framework), which will be a pain (because every plugin will need updating)...  5:S

Are we making use of the None return value in particular?

dgmcdona replied:
Not particularly; I'll revert this.

ikelos commented on volatility3/framework/symbols/windows/extensions/registry.py:
```diff
@@ -132,6 +175,103 @@ def get_full_key_name(self) -> str:
         return "\\".join(reversed(output))
 
 
+class CM_KEY_VALUE(objects.StructType):
```
It;s really hard to see what's changed when this has (mostly) just moved position in the file).  I realize it's been done to allow for typing, but I'd almost sooner split it into two PRs, one where the type is in quotes for delayed resolution, and then another shifting the code about...  5:S

dgmcdona replied:
Moved it back to it's original position 👍

ikelos commented on volatility3/framework/symbols/windows/extensions/registry.py:
```diff
@@ -52,6 +77,24 @@ class RegKeyFlags(enum.IntEnum):
     KEY_VIRTUAL_STORE = 0x200
 
 
+class UnexpectedValueTypeException(exceptions.VolatilityException):
```
I wondering if this shouldn't live in `exceptions` rather than off on its own here?

ikelos commented on volatility3/framework/plugins/windows/amcache.py:
```diff
@@ -0,0 +1,690 @@
+                sha1_hash = sha1_hash[4:] if sha1_hash.startswith("0000") else sha1_hash
+
+            company, product, service, last_write_time, driver_timestamp = (
+                binary_key.get_string_value(
```
This makes for very long-winded code, because we can't use `None` as the default and then change `None` to `UnparseableValue()` using `or UnparseableValue()` even though that's the majority of the cases here (since `None` as the default causes it to throw the exception)...  Not certain what to do here but it feels like having to specify a default value every time is not ideal.

I'm leaning to just returning `None` whenever it can't deal with a value and the people could call `get_value_data` if they needed it to throw the wrong_type exception.  Otherwise the `get_binary_data` methods, etc, are all just glorified ways of catching the exception?

ikelos replied:
That would then allow all the `get_x_data(thing, UnparsableValues())` calls to be swapped out for `or UnparsableValue()` at the time the variable is spat out of the generator.

dgmcdona replied:
I think I was trying too hard to avoid the requirements of the language (handling exceptions) in order to prevent future runtime errors. It's probably better to go back to the way things were before, for the most part, and plugin authors will just have to be careful about checking values types in their own code.

ikelos reviewed (commented):
Hiya, sorry the unparsablevalues are causing such a headache.  Looking better now, but there's a couple changes that would require bumping version numbers (one of them the framework MAJOR version number, so I'd prefer to avoid that if possible).

The `get_x_data()` methods are a little cumbersome, and I think it would be better if they always returned `None` when we're not interested in the `UnexpectedType` exception (and then people can call `get_value_data()` when they actually need the exception.  That might also remove a lot of excess default values that could be handled as more succinctly in the generator return.  Otherwise just a couple bits and pieces, but I do appreciate you taking the time to help me make sure that the API is all good for the future as well as this particular use-case...  5:)

dgmcdona commented:
After thinking things over, I decided to roll back the additional methods that I had originally planned to add to the registry extension `CM_KEY_VALUE`/`CM_KEY_NODE` classes, as they were possibly introducing more complexity than they were solving problems. As far as differentiating on missing vs unparsable values, I have moved that logic into helper functions in the `amcache` module. 

I also removed the changes that would have required a major framework version bump.

ikelos commented on volatility3/framework/symbols/windows/extensions/registry.py:
```diff
@@ -249,6 +282,10 @@ def get_name(self) -> interfaces.objects.ObjectInterface:
         self.Name.count = namelength
         return self.Name.cast("string", max_length=namelength, encoding="latin-1")
 
+    def get_type(self) -> RegValueTypes:
```
Additions like this should still bump the MINOR version number of the framework, but luckily nothing else will have to change (any plugins that call these methods though will have to depend on that later framework version).

ikelos commented on volatility3/framework/plugins/windows/amcache.py:
```diff
@@ -0,0 +1,646 @@
+    if not isinstance(data, int):
+        return renderers.UnparsableValue()
+
+    if isinstance(data, str):
```
This seems... strange?  Does the registry record datetime values as... strings?!?!  Without a timezone?!?!?

dgmcdona replied:
It sure does :) This Windows component encodes timestamps in at least 3 different formats (FILETIME, unix epoch, this string format)

ikelos commented on volatility3/framework/plugins/windows/amcache.py:
```diff
@@ -0,0 +1,646 @@
+            values = {
+                str(value.get_name()): value
+                for value in program_key.get_values()
+                if value.get_name() in [key.value for key in Win10InvAppValName]
```
Almost certainly a small point, but this list comprehension looks like it could be done once outside the loop and then just referenced in the loop?

ikelos commented on volatility3/framework/plugins/windows/amcache.py:
```diff
@@ -0,0 +1,646 @@
+
+            product: str = name if isinstance(name, str) else "UNKNOWN"  # type: ignore
+            if isinstance(version, str):
+                product += " " + version
```
Should we not be returning `version` as a separate field, so another plugin could parse it easily?  I guess they're both strings, it just feels like if they're separate now, why combine them rather than pass them through as two fields?

ikelos commented on volatility3/framework/plugins/windows/amcache.py:
```diff
@@ -0,0 +1,646 @@
+            values = {
+                str(value.get_name()): value
+                for value in file_key.get_values()
+                if value.get_name() in [key.value for key in valName]
```
Again, this list comprehension could be done before the loop even starts rather than every run of a loop inside a loop...

ikelos commented on volatility3/framework/plugins/windows/amcache.py:
```diff
@@ -0,0 +1,646 @@
+            values = {
+                str(value.get_name()): value
+                for value in binary_key.get_values()
+                if value.get_name() in [key.value for key in valName]
```
Same again

ikelos commented on volatility3/framework/plugins/windows/amcache.py:
```diff
@@ -0,0 +1,646 @@
+        """
+        for binary_key in driver_binary_key.get_subkeys():
+
+            valName = Win10DriverBinaryValName
```
This doesn't need to be reset each time round the loop, feels like it could be set once before the loop starts...

ikelos reviewed (commented):
Cool, just gave it another quick once over.  It does need a version bump, but luckily nothing overly dramatic!  Otherwise there's a little optimization I think we could do but it's pretty much good to go.  The version bump (on the framework, and in the plugin `required_framework_version` are all that'd be needed).

dgmcdona committed:
Remove unused method

dgmcdona committed:
Code review fixes

dgmcdona committed:
Minor version bump: registry extensions API addition

dgmcdona committed:
Fix type signature on get_name

dgmcdona committed:
Add product version to output columns

ikelos reviewed (approved):
Awesome, thanks for fixing up those last points!  5:)
